pkg/plugins/gateway/algorithms/vtc/vtc_basic.go

```diff
@@ -0,0 +1,199 @@
+	utilizationWeight = utils.LoadEnvFloat(VTC_UTILIZATION_WEIGHT, defaultUtilizationWeight)
+	maxPodLoad        = utils.LoadEnvFloat(VTC_MAX_POD_LOAD, defaultMaxPodLoad)
+	inputTokenWeight  = utils.LoadEnvFloat(VTC_INPUT_TOKEN_WEIGHT, defaultInputTokenWeight)
+	outputTokenWeight = utils.LoadEnvFloat(VTC_OUTPUT_TOKEN_WEIGHT, defaultOutputTokenWeight)
+)
+
+// BasicVTCRouter implements the VTC routing algorithm
+type BasicVTCRouter struct {
+	cache          cache.Cache
+	tokenTracker   TokenTracker
+	tokenEstimator TokenEstimator
+	config         *VTCConfig
+}
+
+// NewBasicVTCRouter creates a new BasicVTCRouter with the provided token tracker and estimator
+func NewBasicVTCRouter(tokenTracker TokenTracker, tokenEstimator TokenEstimator, config *VTCConfig) (*BasicVTCRouter, error) {
+	c, err := cache.Get()
+	if err != nil {
+		klog.Error("fail to get cache store in basic-vtc router")
+		return nil, err
+	}
+
+	return &BasicVTCRouter{
+		cache:          c,
+		tokenTracker:   tokenTracker,
+		tokenEstimator: tokenEstimator,
+		config:         config,
+	}, nil
+}
+
+// Route implements the VTC routing algorithm
+func (r *BasicVTCRouter) Route(ctx *types.RoutingContext, pods types.PodList) (string, error) {
+	rng := rand.New(rand.NewSource(time.Now().UnixNano()))
+
+	if pods.Len() == 0 {
+		return "", fmt.Errorf("no pods to forward request")
+	}
+
+	readyPods := utils.FilterRoutablePods(pods.All())
+	if len(readyPods) == 0 {
+		return "", fmt.Errorf("no ready pods available for routing")
+	}
+
+	user := ctx.User
+	if user == nil {
+		klog.Warningf("VTC routing not possible: user is nil, falling back to random pod selection")
+		randomPod, err := utils.SelectRandomPod(readyPods, rng.Intn)
+		if err != nil {
+			return "", fmt.Errorf("fallback to random pod selection failed: %w", err)
+		}
+		ctx.SetTargetPod(randomPod)
+		return ctx.TargetAddress(), nil
+	}
+
+	inputTokens := r.tokenEstimator.EstimateInputTokens(ctx.Message)
+	outputTokens := r.tokenEstimator.EstimateOutputTokens(ctx.Message)
+
+	userTokens, err := r.tokenTracker.GetTokenCount(ctx.Context, *user)
+	if err != nil {
+		klog.ErrorS(err, "failed to get user token count, falling back to zero", "user", *user)
+		userTokens = 0
+	}
+
+	klog.InfoS("VTC tokens for user",
+		"user", *user,
+		"tokens", userTokens,
+		"inputTokens", inputTokens,
+		"outputTokens", outputTokens)
+
+	// Select pod based on hybrid VTC approach that balances fairness and utilization
+	// - Fairness: Users with higher token usage get higher-indexed pods
+	// - Utilization: Consider actual pod load to prevent underutilization
+	var targetPod *v1.Pod
+	var minScore float64 = 1<<63 - 1 // Max float64
+
+	// Calculate scores for each pod
+	for i, pod := range readyPods {
+		// 1. Calculate fairness score based on pod index and user tokens
+		// Normalize user tokens to a value between 0 and len(readyPods)-1
+		normalizedTokens := math.Mod(userTokens, float64(len(readyPods)*100)) / 100.0
```

Comment:
Seems here it assumes each 100-token bucket maps to a pod. does it create implicit and possibly unintuitive grouping?

in the case, token 1234 and token 34 have same result, probably not that fair to large token users. the results lacks monotonicity (using more tokens does not necessarily result in being assigned to a later pod). Is that expected in this implementation

Reply:
I agree with your example where this algo creates implicit grouping & lacks monotonicity. Verified using [tests](https://github.com/vllm-project/aibrix/pull/964/commits/280611247ed5f710be4127b7d71d4edf0ee661e7) & this was not intended.

My initial impl fully focussed on fairness component & utilization was not balanced well.  I was experimenting with different algos & modulo impl resulted in good weighted balance in my test scenarios.

But yeah, I'm working on fixing this.

Reply:
Addressed in https://github.com/vllm-project/aibrix/pull/1011